src/vmm/src/device_manager/persist.rs

```diff
@@ -0,0 +1,441 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Provides functionality for saving/restoring the MMIO device manager and its devices.
+
+// Currently only supports x86_64.
+#![cfg(target_arch = "x86_64")]
+
+// TODO: remove once serialization is used.
+#![allow(unused)]
+
+use std::io;
+use std::sync::{Arc, Mutex};
+
+use super::mmio::*;
+
+use devices::virtio::block::persist::{BlockConstructorArgs, BlockState};
+use devices::virtio::block::Block;
+use devices::virtio::net::persist::{Error as NetError, NetConstructorArgs, NetState};
+use devices::virtio::net::Net;
+use devices::virtio::persist::{MmioTransportConstructorArgs, MmioTransportState};
+use devices::virtio::vsock::persist::{VsockConstructorArgs, VsockState, VsockUdsConstructorArgs};
+use devices::virtio::vsock::{Vsock, VsockError, VsockUnixBackend, VsockUnixBackendError};
+use devices::virtio::{MmioTransport, TYPE_BLOCK, TYPE_NET, TYPE_VSOCK};
+use kvm_ioctls::VmFd;
+use snapshot::Persist;
+use versionize::{VersionMap, Versionize, VersionizeResult};
+use versionize_derive::Versionize;
+use vm_memory::GuestMemoryMmap;
+
+/// Errors for (de)serialization of the MMIO device manager.
+#[derive(Debug)]
+pub enum Error {
+    Block(io::Error),
+    DeviceManager(super::mmio::Error),
+    MmioTransport,
+    Net(NetError),
+    Vsock(VsockError),
+    VsockUnixBackend(VsockUnixBackendError),
+}
+
+#[derive(Versionize)]
+/// Holds the state of a block device connected to the MMIO space.
+pub struct ConnectedBlockState {
+    /// Device identifier.
+    pub device_id: String,
+    /// Device state.
+    pub device_state: BlockState,
+    /// Mmio transport state.
+    pub transport_state: MmioTransportState,
+    /// VmmResources.
+    pub mmio_slot: MMIODeviceInfo,
+}
+
+#[derive(Versionize)]
+/// Holds the state of a net device connected to the MMIO space.
+pub struct ConnectedNetState {
+    /// Device identifier.
+    pub device_id: String,
+    /// Device state.
+    pub device_state: NetState,
+    /// Mmio transport state.
+    pub transport_state: MmioTransportState,
+    /// VmmResources.
+    pub mmio_slot: MMIODeviceInfo,
+}
+
+#[derive(Versionize)]
+/// Holds the state of a vsock device connected to the MMIO space.
+pub struct ConnectedVsockState {
+    /// Device identifier.
+    pub device_id: String,
+    /// Device state.
+    pub device_state: VsockState,
+    /// Mmio transport state.
+    pub transport_state: MmioTransportState,
+    /// VmmResources.
+    pub mmio_slot: MMIODeviceInfo,
+}
+
+#[derive(Versionize)]
+/// Holds the device states.
+pub struct DeviceStates {
+    /// Block device states.
+    pub block_devices: Vec<ConnectedBlockState>,
+    /// Net device states.
+    pub net_devices: Vec<ConnectedNetState>,
+    /// Vsock device tests.
+    pub vsock_device: Option<ConnectedVsockState>,
+}
+
+pub struct MMIODevManagerConstructorArgs<'a> {
+    pub mem: GuestMemoryMmap,
+    pub vm: &'a VmFd,
+}
+
+impl<'a> Persist<'a> for MMIODeviceManager {
```

Comment:
This looks nice! One thing that I would worry about here is that the MMIODeviceManager becomes coupled with the actual devices.